x/cronos/keeper/ibc_hooks.go

```diff
@@ -0,0 +1,44 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/modules/apps/transfer/types"
+)
+
+var _ types.TransferHooks = Keeper{}
+
+func (k Keeper) AfterSendTransfer(
+	ctx sdk.Context,
+	sourcePort, sourceChannel string,
+	token sdk.Coin,
+	sender sdk.AccAddress,
+	receiver string,
+	isSource bool) {
+}
+
+func (k Keeper) AfterRecvTransfer(
+	ctx sdk.Context,
+	destPort, destChannel string,
+	token sdk.Coin,
+	receiver string,
+	isSource bool) {
+	// Only after minting vouchers
+	if !isSource {
+		err := k.ConvertVouchersToEvmCoins(ctx, receiver, sdk.NewCoins(token))
+		if err != nil {
```

Comment:
The error is ignored, is there a possibility we end up with a partial state?

Reply:
yeah I though about it.
I think it is better to let the IBC transaction success and eventually let only the hook fails. Otherwise I am not sure user can recover its coins.

Even if hook fails, user will receive native token, he can either
- Send back to source chain to refund through IBC
- Wait bugfix in cronos and retry manually to convert vouchers to evm coins

Reply:
snapshot and revert sounds good too, but not sure how to achieve it

Reply:
> snapshot and revert sounds good too, but not sure how to achieve it. Maybe with cache context?

yes, I was thinking about cache context.
hook failure as a whole is fine, but the partially written state is a disaster.

Reply:
added revert state logic and add unit tests